### ISSUE TITLE
[framework] Inaccessibility of supply_mut and treasury_into_supply via Coin Manager

### ISSUE THREAD

**Issue by miker83z** [iotaledger/iota#4594]
FROM THE AUDIT:

 > ## Summary
> The functions `supply_mut` and `treasury_into_supply` are not accessible through the Coin Manager.
> 
> ## Vulnerability Detail
> The Coin Manager is designed to act as a central shared object, allowing seamless access to all properties of a Coin. However, it fails to expose the `supply_mut` and `treasury_into_supply` functions.
> 
> * `supply_mut`: Allows direct mutability of the Coin’s supply.
> * `treasury_into_supply`: Transfers assets from the treasury into the supply.
> 
> These functions can be accessed and utilized through the Coin module if users choose not to rely on the Coin Manager for managing operations. However, it becomes impossible to call them directly via the manager if they choose to use the Coin Manager.
> 
> ## Impact
> Direct access to the functions `supply_mut` and `treasury_into_supply` is not possible through the Coin Manager.
> 
> ## Code Snippet
> ```move
>     // === Supply <-> TreasuryCap morphing and accessors  ===
> 
>     /// Unwrap `TreasuryCap` getting the `Supply`.
>     ///
>     /// Operation is irreversible. Supply cannot be converted into a `TreasuryCap` due
>     /// to different security guarantees (TreasuryCap can be created only once for a type)
>     
>     public fun treasury_into_supply<T>(treasury: TreasuryCap<T>): Supply<T> {
>         let TreasuryCap { id, total_supply } = treasury;
>         id.delete();
>         total_supply
>     }
> 
>     /// Get mutable reference to the treasury's `Supply`.
>     public fun supply_mut<T>(treasury: &mut TreasuryCap<T>): &mut Supply<T> {
>         &mut treasury.total_supply
>     }
> ```
> 
> ## Tool used
> Manual Review
> 
> ## Recommendation
> Add `supply_mut` and `treasury_into_supply` as accessible methods within the Coin Manager.

**Comment by valeriyr** [comment]
@miker83z @lzpap Currently, the `CoinManager` is designed to manage the supply internally by applying additional checks like `maximum_supply`, for example. Adding these functions will affect the logic, so we don't need to add them as is, I think.

**Comment by lzpap** [comment]
> [@miker83z](https://github.com/miker83z) [@lzpap](https://github.com/lzpap) Currently, the `CoinManager` is designed to manage the supply internally by applying additional checks like `maximum_supply`, for example. Adding these functions will affect the logic, so we don't need to add them as is, I think.

Indeed, exposing the functions negates the logic of the `CoinManager`, that is, `maximum_supply` can not be enforced anymore.

**Comment by valeriyr** [comment]
To add the `treasury_into_supply` function we need to destroy the related `TreasuryCap` that is not supported, as well as `CoinManager` destroying.

`supply_mut` can't be added because the total supply should be handled internally to support the maximum total supply feature:

```rust
#[test]
fun test_mutable_supply() {
    let sender = @0xA;
    let mut scenario = test_scenario::begin(sender);
    let witness = COIN_MANAGER_TESTS{};

    // Create a `Coin`.
    let (cap, meta) = coin::create_currency(
        witness,
        0, 
        b"TEST",
        b"TEST",
        b"TEST",
        option::none(),
        scenario.ctx(),
    );

    let (cmcap, metacap, mut wrapper) = coin_manager::new(cap, meta, scenario.ctx());
    
    // We should start out with a supply of 0.
    assert_eq(wrapper.total_supply(), 0);

    // Enforce a maximum supply.
    cmcap.enforce_maximum_supply(&mut wrapper, 10);
    
    // Mint a balance directly using a mutable supply.
    let balance = wrapper.supply_mut(&cmcap).increase_supply(100);

    // Check that the CoinManager instance is in an inconsistent state.
    assert_eq(wrapper.total_supply(), 100);
    assert_eq(wrapper.maximum_supply(), 10);

    transfer::public_transfer(balance.into_coin(scenario.ctx()), sender);
    transfer::public_transfer(cmcap, sender);
    metacap.renounce_metadata_ownership(&mut wrapper);

    transfer::public_share_object(wrapper);

    scenario.end();
}
```
